service/handlers.go

```diff
@@ -1210,20 +1219,55 @@ func (h handler) listTokens(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	tokens, err := h.dbClient.ListTokenEntries(ctx, projectName)
+	dbTokens, err := h.dbClient.ListTokenEntries(ctx, projectName)
+	if err != nil {
+		level.Error(l).Log("message", "error listing project tokens", "error", err)
+		h.errorResponse(w, "error listing project tokens", http.StatusInternalServerError)
+		return
+	}
+
+	cpTokens, err := cp.ListProjectTokens(projectName)
 	if err != nil {
 		level.Error(l).Log("message", "error listing project tokens", "error", err)
 		h.errorResponse(w, "error listing project tokens", http.StatusInternalServerError)
 		return
 	}
 
 	resp := []responses.ListTokens{}
-	for _, tokenEntry := range tokens {
-		resp = append(resp, responses.ListTokens{
-			CreatedAt: tokenEntry.CreatedAt,
-			ExpiresAt: tokenEntry.ExpiresAt,
-			TokenID:   tokenEntry.TokenID,
-		})
+
+	// Vault automatically deletes secrets when they expire, so there is a chance that the tokens DB and Vault
+	// are not in sync.
+	//
+	// Check if the two project tokens list are equal. If not equal, return CP tokens as the CP
+	// is the source of truth.
+	// Else, return the DB tokens as they already contain the expire time.
+	//
+	// This is assuming that the expire time for the DB token is correct since the token exists in the CP.
+	if isProjectTokensListsEqual(cpTokens, dbTokens) {
+		for _, tokenEntry := range dbTokens {
```

Comment:
I'm thinking this for loop can be put into a handler function, maybe something like `toListTokensResponse(dbTokens)`. What do you guys think?